File: task-manager-ios/task-manager-ios/Data/Repositories/TaskRepositoryImpl.swift
//
//  TaskRepositoryImpl.swift
//  task-manager-ios
//
//  Created by Jeofferson Dela Peña on 3/9/24.
//

import Foundation
import RealmSwift

class TaskRepositoryImpl: TaskRepository {
    private let remoteService: TaskRemoteService
    private let localService: TaskLocalService
    
    init(remoteService: TaskRemoteService, localService: TaskLocalService) {
        self.remoteService = remoteService
        self.localService = localService
    }
    
    func getTasks() -> Result<[Task], TaskManagerError> {
        if /* !hasInternetConnection */ true {
            switch localService.getTasks() {
            case .success(let taskLocalDTOs):
                return .success(taskLocalDTOs.toDomain())
            case .failure(let error):
                return .failure(error)
            }
        } else {
            switch remoteService.getTasks() {
            case .success(let taskRemoteDTOs):
                cacheTasks(taskRemoteDTOs)
                return .success(taskRemoteDTOs.toDomain())
            case .failure(let error):
                return .failure(error)
            }
        }
    }
    
    func addTask(_ task: Task) -> Result<Void, TaskManagerError> {
        if /* !hasInternetConnection */ true {
            return localService.addTask(TaskLocalDTO.toData(task))
        } else {
            return remoteService.addTask(TaskRemoteDTO.toData(task))
        }
    }
    
    func editTask(_ task: Task) -> Result<Void, TaskManagerError> {
        if /* !hasInternetConnection */ true {
            return localService.editTask(TaskLocalDTO.toData(task))
        } else {
            return remoteService.editTask(TaskRemoteDTO.toData(task))
        }
    }
    
    func toggleTaskCompletion(_ task: Task) -> Result<Void, TaskManagerError> {
        if /* !hasInternetConnection */ true {
            return localService.toggleTaskCompletion(TaskLocalDTO.toData(task))
        } else {
            return remoteService.toggleTaskCompletion(TaskRemoteDTO.toData(task))
        }
    }
    
    func deleteTask(_ task: Task) -> Result<Void, TaskManagerError> {
        if /* !hasInternetConnection */ true {
            return localService.deleteTask(TaskLocalDTO.toData(task))
        } else {
            return remoteService.deleteTask(TaskRemoteDTO.toData(task))
        }
    }
    
    private func cacheTasks(_ taskRemoteDTOs: [TaskRemoteDTO]) {
        _ = localService.addTask(
            TaskLocalDTO.toData(taskRemoteDTOs.toDomain())
        )
    }
}

File: task-manager-android/app/src/main/java/com/jeoffersondelapena/task_manager_android/data/repository/TaskRepositoryImpl.kt
package com.jeoffersondelapena.task_manager_android.data.repository

import com.jeoffersondelapena.task_manager_android.data.local.dto.TaskLocalDto
import com.jeoffersondelapena.task_manager_android.data.local.dto.toDomain
import com.jeoffersondelapena.task_manager_android.data.local.service.TaskLocalService
import com.jeoffersondelapena.task_manager_android.data.remote.dto.TaskRemoteDto
import com.jeoffersondelapena.task_manager_android.data.remote.dto.toDomain
import com.jeoffersondelapena.task_manager_android.data.remote.service.TaskRemoteService
import com.jeoffersondelapena.task_manager_android.domain.model.Task
import com.jeoffersondelapena.task_manager_android.domain.repository.TaskRepository
import com.jeoffersondelapena.task_manager_android.domain.util.helper.TaskManagerException
import com.jeoffersondelapena.task_manager_android.domain.util.helper.TaskManagerResult
import javax.inject.Inject

class TaskRepositoryImpl @Inject constructor(
    private val taskRemoteService: TaskRemoteService,
    private val taskLocalService: TaskLocalService,
) : TaskRepository {
    override fun getTasks(): TaskManagerResult<List<Task>, TaskManagerException> {
        if (/* !hasInternetConnection */ true) {
            return when (val result = taskLocalService.getTasks()) {
                is TaskManagerResult.Success -> {
                    TaskManagerResult.Success(result.value.toDomain())
                }
                is TaskManagerResult.Failure -> {
                    TaskManagerResult.Failure(result.exception)
                }
            }
        } else {
            return when (val result = taskRemoteService.getTasks()) {
                is TaskManagerResult.Success -> {
                    val taskRemoteDtos = result.value
                    cacheTasks(taskRemoteDtos)
                    TaskManagerResult.Success(taskRemoteDtos.toDomain())
                }
                is TaskManagerResult.Failure -> {
                    TaskManagerResult.Failure(result.exception)
                }
            }
        }
    }

    override fun addTask(task: Task): TaskManagerResult<Unit, TaskManagerException> {
        if (/* !hasInternetConnection */ true) {
            return taskLocalService.addTask(TaskLocalDto.toData(task))
        } else {
            return taskRemoteService.addTask(TaskRemoteDto.toData(task))
        }
    }

    override fun editTask(task: Task): TaskManagerResult<Unit, TaskManagerException> {
        if (/* !hasInternetConnection */ true) {
            return taskLocalService.editTask(TaskLocalDto.toData(task))
        } else {
            return taskRemoteService.editTask(TaskRemoteDto.toData(task))
        }
    }

    override fun toggleTaskCompletion(task: Task): TaskManagerResult<Unit, TaskManagerException> {
        if (/* !hasInternetConnection */ true) {
            return taskLocalService.toggleTaskCompletion(TaskLocalDto.toData(task))
        } else {
            return taskRemoteService.toggleTaskCompletion(TaskRemoteDto.toData(task))
        }
    }

    override fun deleteTask(task: Task): TaskManagerResult<Unit, TaskManagerException> {
        if (/* !hasInternetConnection */ true) {
            return taskLocalService.deleteTask(TaskLocalDto.toData(task))
        } else {
            return taskRemoteService.deleteTask(TaskRemoteDto.toData(task))
        }
    }

    private fun cacheTasks(taskRemoteDtos: List<TaskRemoteDto>) {
        taskLocalService.addTask(TaskLocalDto.toData(taskRemoteDtos.toDomain()))
    }
}
